inference-engine/thirdparty/clDNN/runtime/engine.cpp

```diff
@@ -120,22 +120,49 @@ memory_ptr engine::share_surface(const layout& layout, shared_surface surf, uint
 #endif  // _WIN32
 
 uint64_t engine::get_max_used_device_memory() const {
-    return peak_memory_usage.load();
+    uint64_t total_peak_memory_usage {0};
+    for (auto const& m : peak_memory_usage_map) {
+        total_peak_memory_usage += m.second.load();
+    }
+    return total_peak_memory_usage;
 }
 
-uint64_t engine::get_used_device_memory() const {
-    return memory_usage.load();
+uint64_t engine::get_max_used_device_memory(allocation_type type) const {
+    uint64_t peak_memory_usage {0};
+    auto iter = peak_memory_usage_map.find(type);
+    if (iter != peak_memory_usage_map.end()) {
+        peak_memory_usage = iter->second.load();
+    }
+    return peak_memory_usage;
 }
 
-void engine::add_memory_used(size_t bytes) {
-    memory_usage += bytes;
-    if (memory_usage > peak_memory_usage) {
-        peak_memory_usage = memory_usage.load();
+uint64_t engine::get_used_device_memory(allocation_type type) const {
+    uint64_t memory_usage {0};
+    auto iter = memory_usage_map.find(type);
+    if (iter != memory_usage_map.end()) {
+        memory_usage = iter->second.load();
     }
+    return memory_usage;
 }
 
-void engine::subtract_memory_used(size_t bytes) {
-    memory_usage -= bytes;
+void engine::add_memory_used(size_t bytes, allocation_type type) {
+    if (!memory_usage_map.count(type) && !peak_memory_usage_map.count(type)) {
+        memory_usage_map.insert(std::make_pair(type, 0));
+        peak_memory_usage_map.insert(std::make_pair(type, 0));
```

Comment:
This part is not thread safe. Please check add/subtract impls to ensure thread safety of all operations.

Reply:
done

Reply:
@andrew-k-park  Could you try multi-threaded application by yourself? e.g., fix benchmark app to run multiple threads inside it

Reply:
As we discussed, I've confirmed thread safety when I try to run multi-threaded sample app, memory map works well as intended.